File: LeetCodeKotlin/SqrtX_69.kt
class SqrtX_69 {
    fun mySqrt(x: Int): Int {
        // Handle edge cases directly
        if (x < 2) return x

        var left = 1
        var right = x / 2  // For x>=2, the square root is always <= x/2

        // Binary search for the integer square root
        while (left <= right) {
            val mid = left + (right - left) / 2
            // Instead of mid*mid, use mid <= x/mid to avoid overflow.
            if (mid <= x / mid) {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }

        return right
    }
}

File: LeetCodeCpp/SqrtX_69.cpp
#include "SqrtX_69.h"

// Binary search for the last value whose square does not exceed x.
// See Docs/Algorithms/Binary_Search.md for the "last true" pattern.
int SqrtX_69::mySqrt(int x)
{
    if (x < 2) {
        return x;
    }

    int left = 1;
    int right = x / 2;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (mid <= x / mid) { // mid * mid <= x, without overflow.
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return right;
}
